java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java

```diff
@@ -0,0 +1,415 @@
+import com.google.common.collect.Maps;
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+import org.sonar.java.se.ConstraintManager.NullConstraint;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public interface SymbolicValue {
+
+  SymbolicValue NULL_LITERAL = new ObjectSymbolicValue(0) {
+    @Override
+    public String toString() {
+      return super.toString()+"_NULL";
+    }
+  };
+  SymbolicValue TRUE_LITERAL = new ObjectSymbolicValue(1){
+    @Override
+    public String toString() {
+      return super.toString()+"_TRUE";
+    }
+  };
+  SymbolicValue FALSE_LITERAL = new ObjectSymbolicValue(2){
+    @Override
+    public String toString() {
+      return super.toString()+"_FALSE";
+    }
+  };
+
+  void computedFrom(List<SymbolicValue> symbolicValues);
+
+  ProgramState setSingleConstraint(ProgramState state, NullConstraint notNull);
+
+  List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint);
+
+  List<ProgramState> setConstraint(ProgramState programState, NullConstraint nullConstraint);
+
+  class ObjectSymbolicValue implements SymbolicValue {
+
+    private final int id;
+
+    public ObjectSymbolicValue(int id) {
+      this.id = id;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      ObjectSymbolicValue that = (ObjectSymbolicValue) o;
+      return Objects.equal(id, that.id);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(id);
+    }
+
+    @Override
+    public String toString() {
+      return "SV#" + id;
+    }
+
+    @Override
+    public void computedFrom(List<SymbolicValue> symbolicValues) {
+      // no op in general case
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, NullConstraint nullConstraint) {
+      Object data = programState.constraints.get(this);
+      if (data instanceof NullConstraint) {
+        NullConstraint nc = (NullConstraint) data;
+        if ((NullConstraint.NULL.equals(nullConstraint) && NullConstraint.NOT_NULL.equals(nc)) ||
+          (NullConstraint.NULL.equals(nc) && NullConstraint.NOT_NULL.equals(nullConstraint))) {
+          // setting null where value is known to be non null or the contrary
+          return ImmutableList.of();
+        }
+      }
+      if (data == null || !data.equals(nullConstraint)) {
+        Map<SymbolicValue, Object> temp = Maps.newHashMap(programState.constraints);
+        temp.put(this, nullConstraint);
+        return ImmutableList.of(new ProgramState(programState.values, temp, programState.visitedPoints, programState.stack));
+      }
+      return ImmutableList.of(programState);
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "setConstraint" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)